### PR TITLE
feat: async compile in lsp

### PR DESCRIPTION
**Pull request by He1pa**
feat: asymc compile in lsp 
1 support async compile when vfs changes(open file and ChangeTextDocument). 2. remove scope in lsp db

#### 1. Does this PR affect any open issues?(Y/N) and add issue references (e.g. "fix #123", "re #123".):
lsp
- [ ] N
- [x] Y #915  

#### 3. Provide a description of the PR(e.g. more details, effects, motivations or doc link):

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [x] Other

#### 4. Are there any breaking changes?(Y/N) and describe the breaking changes(e.g. more details, motivations or doc link):

- [x] N
- [ ] Y 

#### 5. Are there test cases for these changes?(Y/N) select and add more details, references or doc links:

- [x] Unit test
- [ ] Integration test
- [ ] Benchmark (add benchmark stats below)
- [ ] Manual test (add detailed scripts or steps below)
- [ ] Other

### CODE REVIEW

File: kclvm/tools/src/LSP/src/state.rs [hunk 1]
```diff
@@ -208,28 +199,48 @@ impl LanguageServerState {
...
-                            self.analysis.set_db(
-                                file.file_id,
-                                AnalysisDatabase {
-                                    prog,
-                                    scope,
-                                    diags,
-                                    gs,
-                                },
-                            );
+                    let mut snapshot = self.snapshot();
```
Comment (Peefy) [inline kclvm/tools/src/LSP/src/state.rs:202]: remove the dup code.

File: kclvm/tools/src/LSP/src/state.rs [hunk 2]
```diff
@@ -208,28 +199,48 @@ impl LanguageServerState {
...
+
+                    self.thread_pool.execute({
+                        let mut snapshot = self.snapshot();
+                        let sender = self.task_sender.clone();
+                        let module_cache = self.module_cache.clone();
+                        let uri = url(&snapshot, file.file_id).unwrap();
```
Comment (Peefy) [inline kclvm/tools/src/LSP/src/state.rs:208]: prevent the `unwrap()` code.

File: kclvm/tools/src/LSP/src/state.rs [hunk 3]
```diff
@@ -208,28 +199,48 @@ impl LanguageServerState {
...
+                                    .flat_map(|diag| kcl_diag_to_lsp_diags(diag, filename.as_str()))
+                                    .collect::<Vec<Diagnostic>>();
+                                sender.send(Task::Notify(lsp_server::Notification {
+                                    method: PublishDiagnostics::METHOD.to_owned(),
+                                    params: to_json(PublishDiagnosticsParams {
+                                        uri,
+                                        diagnostics,
+                                        version: None,
+                                    })
+                                    .unwrap(),
```
Comment (Peefy) [inline kclvm/tools/src/LSP/src/state.rs:238]: prevent the `unwrap()` code.

**Comment by Peefy** [comment]
@CodiumAI-Agent /review

File: kclvm/tools/src/LSP/src/state.rs (hunk 3 above)
Comment (He1pa) [reply kclvm/tools/src/LSP/src/state.rs:238]: It is clearly safe here， this unwrap() as same as https://github.com/kcl-lang/kcl/issues/760 and https://github.com/kcl-lang/kcl/pull/778

File: kclvm/tools/src/LSP/src/request.rs [hunk 4]
```diff
@@ -202,22 +203,38 @@ pub(crate) fn handle_completion(
...
+        Some(ch) => match ch {
+            '=' | ':' => {
+                let db = parse_param_and_compile(
+                    Param {
+                        file: file.clone(),
+                        module_cache: snapshot.module_cache.clone(),
+                    },
+                    Some(snapshot.vfs.clone()),
+                )
+                .unwrap();
```
Comment (Peefy) [inline kclvm/tools/src/LSP/src/request.rs:222]: Prevent the `unwrap()` code.

File: kclvm/tools/src/LSP/src/request.rs [hunk 5]
```diff
@@ -253,11 +270,22 @@ pub(crate) fn handle_document_symbol(
...
-    let res = document_symbol(&file, &db.gs);
+
+    let db = parse_param_and_compile(
+        Param {
+            file: file.clone(),
+            module_cache: snapshot.module_cache.clone(),
+        },
+        Some(snapshot.vfs.clone()),
+    )
+    .unwrap();
```
Comment (Peefy) [inline kclvm/tools/src/LSP/src/request.rs:286]: Prevent the `unwrap()` code.

**Comment by Peefy** [comment]
@CodiumAI-Agent /review

File: kclvm/tools/src/LSP/src/analysis.rs [hunk 6]
```diff
@@ -1,14 +1,9 @@
 use crate::db::AnalysisDatabase;
```
Comment (Peefy) [inline kclvm/tools/src/LSP/src/analysis.rs:1]: PR confilct

**Review by Peefy — APPROVED** [review]
LGTM

**Pull request merged**